### ISSUE TITLE
builder doesn't build

### ISSUE THREAD

**Issue by Spongman** [ldtteam/minecolonies#6236]
### Prerequisites 

- [ ] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [ ] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [ ] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.15.2
- MineColonies Version: minecolonies-0.13.438-ALPHA-universal
- Structurize Version: structurize-0.13.94-ALPHA-universal
- Related Mods and their Versions: Craft to Exile modpack

## Expected behavior
builder builds

## Actual behavior
builder doesn't build, just wanders around doing nothing. it built a Builder's Hut, but refuses to build the forester's hut. I clicked 'Build Building', but no worker is assigned:

NOTE: "Worker Assigned:"

![image](https://user-images.githubusercontent.com/1088194/101982196-fd3d2a00-3c26-11eb-9170-9a12da8db80d.png)

![image](https://user-images.githubusercontent.com/1088194/101982109-899b1d00-3c26-11eb-9086-d4064ea8e343.png)

## Steps to reproduce the issue

I have no idea. This mod is so confusing, it's never apparent whether or not your actions in the UI actually have any effect - the citizens are so slow to react (if at all) that there's no indication that what they're doing is because of something you told them to do, or just them wandering around on their own.
maybe it would be less confusing if you reduced the time between UI actions and feedback. i understand having the builder take a while to build the building, but if I hired a builder IRL and i asked him to do go and a job somewhere and two weeks later he came back and said 'i need a hammer' he'd be fired on the spot.

**Comment by Raycoms** [comment]
Please update, you are a bunch of versions behind. The reason the workers are not "directly responsive" is such that the mod doesn't bring down the server even though you can get hundreds of citizens. Workers at low levels are also significantly slower than workers at higher levels.

**Comment by Raycoms** [comment]
If the problem persists after updating please do some trouble shooting in the discord. Workers don't work during rain for example either.

**Comment by Spongman** [comment]
> The reason the workers are not "directly responsive" is such that the mod doesn't bring down the server even though you can get hundreds of citizens

i'm not sure i understand this. i can have hundreds of chests on a server and yet when i open one of them it instantly shows me its contents (not a client-openly operation).


how do i update? i think i have the latest version of this modpack.

sorry, i'm not going to use discord. i hate the fact that all messages are in a single thread- it's impossible to follow anything on there.

**Comment by ravenbuilder934** [comment]
> i'm not sure i understand this. i can have hundreds of chests on a server and yet when i open one of them it instantly shows me its contents (not a client-openly operation).

Citizens are a lot more complicated than chests 😄 They have AI, pathfinding, etc.

> how do i update? i think i have the latest version of this modpack.

You need to update the mod within the modpack. How you'll do that depends on your launcher-which one will you use?

> sorry, i'm not going to use discord. i hate the fact that all messages are in a single thread- it's impossible to follow anything on there.

That's incorrect actually. Discord has multiple channels for different topics, and the reply feature was recently added, which lets you have multiple distinct threads in the same channel.

**Comment by Spongman** [comment]
> Citizens are a lot more complicated than chests

yeah, i understand that, i'm not expecting the builder to immediately go off and start building, but i do expect some kind of feedback that my instructions have at least been heard and whether or not there's currently something preventing them from being carried out. there could even be a tunable cooldown period if there was concern for server load - although i'm playing single-player so i don't care about that.

but really, if this wasn't part of a quest progression in this modpack, i would have given up on this frustrating mess a long time ago. there's simply just too much standing around waiting for a mob to tell me there's some issue. it doesn't help that there's a whole slew of other issues: the inventory's always full (it isn't), the citizens won't eat the food i give them, they need grass blocks i can't get. i'd recommend you do some usability testing - like watch over the shoulders of someone who knows nothing about this mod play it from the start.

**Comment by Raycoms** [comment]
This mod is not for people that expect instant gratification. It takes time to do things and responses are not always immediately. Things take time and you might spend months to build enormous colonies (as we show in our streams and videos as well). We do know that there are certain difficulties for people to start out, especially those that are used to everything happening instantly in all the tech mods. We're addressing those things slowly as well. We don't get money for coding this so we develop the mods primarily for us and our users. If people are interested in playing it they can do the effort to dig deeper and find out how it works and what it has to offer. However, the builder barely takes a walk to the build site and back to their hut to tell you what they need. Their work hut lists all the materials they require. And their interfaces the vast majority of the times tell you what they are up to.

**Comment by Spongman** [comment]
ok, i'm just offering feeback. if you ever wondered, for example, why new players may not be sticking around, please read my comments above again. otherwise, carry on.

**Comment by Raycoms** [comment]
There is a specific target group of new players that stick around for forever after discovering our mod. Maybe there are certain target groups we don't want to stick around :wink:

**Comment by Spongman** [comment]
also, i'm not asking for instant gratification, and i'll ignore the implied insult.

i'm asking for more immediate indication of an error condition.

**Comment by Raycoms** [comment]
Short question too, do you have optifine installed?

**Comment by Raycoms** [comment]
and in 99% of the cases there is an immediate indication

**Comment by Spongman** [comment]
> Short question too, do you have optifine installed?

no, just the vanilla modpack as shipped, plus minecolonies-0.13.479-BETA-universal.

**Comment by ravenbuilder934** [comment]
479 is still out of date, that might be the cause of some of your issues

**Comment by Moleculor** [comment]
> there could even be a tunable cooldown period if there was concern for server load - although i'm playing single-player so i don't care about that.

As someone who *does not work on the mod*, so take this as the uneducated information that might be incorrect that it is, my understanding is that the AI in here actually already is rate-limited to be kind to servers. Which is why you don't see an *instantaneous* response from them. However, in my experience, as long as it's not raining, they aren't sick, hungry, etc (all of which would be visible on their stats page, I believe), they should get to work within about 60 seconds or so.

Also, the very first screenshot you provided with "Worker Assigned:"? Is the space where a lumberjack will be listed, not a builder, after the hut is built. 

You should get a chat message about the construction starting when the builder starts working on it. Specifically when the builder arrives at their builder hut with the intent to start the job on whatever hut (forester, in this case) you want built.

You can see the progress of that build on a Resource Scroll you've assigned to the builder hut block, and should also consider making a Clipboard to make your job easier.

So *if* you have a builder, and by that I mean you have a colonist assigned/hired at the builder's hut, and you've placed a build order for the forester (your screenshot suggests that you have, since it says 'Cancel Build'), then within about a minute (assuming it's not raining) your builder should go to his builder hut and announce he's beginning his build.

Also, you're **~~seventy-seven~~ thirty-six versions behind**. Update. **If** this is a bug, it's likely one that has already been patched.

**Comment by ravenbuilder934** [comment]
They did update, but only to 479 beta, which is still over 30 versions out of date

**Comment by Moleculor** [comment]
To be extra helpful: [Here](https://www.curseforge.com/minecraft/mc-mods/minecolonies/files/all) is where you can find the most recent version. As of the moment I'm writing this it's version 515, but you should grab *whatever* is at the top of that list when you see the link.

**Comment by Spongman** [comment]
> 479 is still out of date, that might be the cause of some of your issues

The 'Download' button on https://www.minecolonies.com/ links to the Curse page https://www.curseforge.com/minecraft/mc-mods/minecolonies. The Download button on there 
![image](https://user-images.githubusercontent.com/1088194/102026948-e2b59e80-3d55-11eb-8701-0e4e09533fec.png)
links to 0.13.437-RELEASE-universal, which is what's included in my modpack.
the most recent version i could find on the 'Files' page of the Curse site is 479 beta:
![image](https://user-images.githubusercontent.com/1088194/102026992-26100d00-3d56-11eb-8899-9923fb9c436d.png)

**Comment by ravenbuilder934** [comment]
You need to click on View All, which will take you to the link Moleculor sent. Then you can see the alphas (recommended update channel)

**Comment by Spongman** [comment]
>  click on View All

ah, behind the 'Beware of the Leopard' sign?

**Comment by Raycoms** [comment]
Curseforge started hiding alphas, not us

**Comment by ravenbuilder934** [comment]
*shrug* Curse put them there, not us

**Comment by ravenbuilder934** [comment]
sniped lol

**Comment by Spongman** [comment]
maybe switch to hosting the builds on github, then you can link to them directly from your site?

**Comment by ravenbuilder934** [comment]
...They're already on github. That's what this repository is.

**Comment by Spongman** [comment]
> ...They're already on github. That's what this repository is.

really? where are the builds on github? i see zipped sources on the releases/tags page, but not binaries.

**Comment by ravenbuilder934** [comment]
> > ...They're already on github. That's what this repository is.
> 
> really? where are the builds on github? i see zipped sources on the releases/tags page, but not binaries.

If you scroll up and go to the Code tab, you'll see *all* the code for the entire mod.

**Comment by Spongman** [comment]
> If you scroll up and go to the Code tab, you'll see _all_ the code for the entire mod.

please read before typing: i said "where are the builds", not "where is the source".

**Comment by ravenbuilder934** [comment]
You said "binaries" not builds. I don't know what you meant by that, but the .jar files (builds) are only on curseforge (for end users).

**Comment by Spongman** [comment]
yeah, and they're hard to find on curseforge, as i pointed out above, and you agreed with me. hence my comment "maybe switch to hosting the builds on github", to which you answered (incorrectly) "They're already on github".

also: https://en.wikipedia.org/wiki/Software_build

**Comment by ravenbuilder934** [comment]
Most people use Curseforge to find mods, as it's a reliable and well-known site (that's the price to pay for bad UI). Putting the builds *only* on github would probably greatly reduce the number of users

**Comment by ravenbuilder934** [comment]
And nowhere on that Wikipedia page does it say anything about "binaries"

**Comment by Spongman** [comment]
binaries, build. same thing.

**Comment by ravenbuilder934** [comment]
Well I've never heard it called that ¯\_(ツ)_/¯

**Comment by Spongman** [comment]
if curseforge is so reliable, why are they hiding your alphas on a separate page behind a hard to find link?

**Comment by ravenbuilder934** [comment]
Reliable != easy-to-use UI. Some third party launchers like https://gdevs.io can help (you can set that one to automatically install alphas, and you can click on the file name to see all classifications *including* alphas)

**Comment by gattsuru** [comment]
Github's also kinda awkward for the level of builds we'd be talking about here, especially since there are two major versions still under simultaneous support.  Doable, and with Actions there's a lot of crazy things you can do!  But not trivial, either.  And people who can deal with github's assets interface can figure out how to use the View All button on curseforge anyway.

It's not that CurseForge is necessarily reliable, especially given Overwolf, so much as they're a trusted and authoritative host for modern modded Minecraft and, unless FTB pulls a magic trick, we're stuck supporting them anyway because of the launcher ecosystem.

As for your builder situation: try the update, make sure to get Structurize to the right version when doing so, make sure your builder has tools matching their current hut level, and that they've got some food on them.

If the problem persists, make sure your serverconfig doesn't have updaterate set sky high in the minecolonies-server.toml file.  A few modpack makers went nuts with that thing.  For solo play, there's basically no reason to set it above 1.

**Comment by Spongman** [comment]
the irony is that you could be CI-ing your builds directly into github releases with some trivial config. it would probably take less time to do than the time it took to fail to understand my comments.

**Comment by Spongman** [comment]
Also, if the first thing you say to any bug report is "upgrade to the latest alpha", then surely the latest alpha is what should be downloaded when you click the 'Download' button, wherever that is?

**Comment by ravenbuilder934** [comment]
There is no "Download" button, there are download buttons for each individual file but not an overall one

**Comment by Spongman** [comment]
> There is no "Download" button, there are download buttons for each individual file but not an overall one

err...

on your site there's this button:
![image](https://user-images.githubusercontent.com/1088194/103390035-1834e880-4ac7-11eb-894e-19517b855ee0.png)

and that takes you to this (really confusing popup, that's a whole other bug):
![image](https://user-images.githubusercontent.com/1088194/103390068-461a2d00-4ac7-11eb-966b-746079949ed5.png)

which takes you to https://www.curseforge.com/minecraft/mc-mods/minecolonies:

![image](https://user-images.githubusercontent.com/1088194/103390105-719d1780-4ac7-11eb-97df-d89ce24265f8.png)


and that link is `minecolonies-0.13.437-RELEASE-universal.jar`

**Comment by ravenbuilder934** [comment]
Blame Curse for making that last download button point to the release, but I'll see what I can do about the one on our site. (https://www.curseforge.com/minecraft/mc-mods/minecolonies also has a link to the latest alpha in the sidebar iirc.)

**Comment by Spongman** [comment]
but that's my whole point. if curse is deserving of blame, then maybe it's time to rethink using it as the destination for your download link? maybe this:
- Ci your builds, tag beta/releases as necessary.
- push a github pages site containing automatically updating links/redirects to the latest builds of each type.
- link to those redirects on your main site.
- skip curse.

**Comment by ravenbuilder934** [comment]
Again. Many people use curse to discover mods. We would be losing all those users.

**Comment by Spongman** [comment]
i'm not saying don't have your mod on curse. i'm just saying don't rely on curse to provide helpful links to the builds you actually want people to run - because it doesn't. there are much better tools for that.

**Comment by gattsuru** [comment]
You literally wrote "skip curse".  If we want people to use alphas in curseforge-downloaded (and, by result, GDLauncher/etc) modpacks, that's not an option.   Might not be what you meant, but not helping your argument.  Even running two build systems in parallel runs into conflict with the DRY principle.

More broadly, this seems a more complicated and less good solution than linking from the web page to the curseforge download all page that lists the alphas instead of an interstitial, _and isn't telling us whether the new version, or the config setting, is related to your problem_.

**Comment by Asherslab** [comment]
> maybe switch to hosting the builds on github, then you can link to them directly from your site?

Curseforge has a lot of major benefits over github or elsewhere. 
A) It's the trusted mod repository for the minecraft modding community
B) Authors (including us) get rewards for the amount of downloads we get, while small it is quite helpful in a situation where we don't get paid directly in any form.
C) Curseforge modpacks cannot use mod files from outside the curseforge platform, hosting our files elsewhere removes us from all those modpacks.
D) Because of the above reasons ***we don't want them on GitHub at all***, among the above it's also simply bound to cause more confusion 

> if curseforge is so reliable, why are they hiding your alphas on a separate page behind a hard to find link?

We've talked with curseforge about this in the past, it relates to the fact that in most cases alphas are considered unstable and therefore not-recommended. we use alphas in a different way. we also talked to them about possibly adding a configuration value regarding allowing alphas to not be hidden on certain projects and they were receptive of the idea. however of course they are currently going through a lot of work with a major transition so that may take some time

> the irony is that you could be CI-ing your builds directly into github releases with some trivial config. it would probably take less time to do than the time it took to fail to understand my comments.

We use CI/CD yes, and yes we could do that relatively easily (although not as easy as you'd think). however i point you back to response 1.

> Also, if the first thing you say to any bug report is "upgrade to the latest alpha", then surely the latest alpha is what should be downloaded when you click the 'Download' button, wherever that is?

pointing back to response 2.


Now. i'm quite tired of this fruitless conversation, this repository is not the location for such a thread or discussion. if you wish to discuss it further you're free to join our discord as has already been stated earlier.
